ambry-rest/src/main/java/com.github.ambry.rest/StorageServerNettyChannelInitializer.java

```diff
@@ -0,0 +1,89 @@
+  private final NettyMetrics nettyMetrics;
+  private final ConnectionStatsHandler connectionStatsHandler;
+  private final RestRequestHandler requestHandler;
+  private final RestServerState restServerState;
+  private final SSLFactory sslFactory;
+
+  /**
+   * Construct a {@link StorageServerNettyChannelInitializer}.
+   * @param nettyConfig the config to use when instantiating certain handlers on this pipeline.
+   * @param performanceConfig the config to use when evaluating ambry service level objectives that include latency.
+   * @param nettyMetrics the {@link NettyMetrics} object to use.
+   * @param connectionStatsHandler the {@link ConnectionStatsHandler} to use.
+   * @param requestHandler the {@link RestRequestHandler} to handle requests on this pipeline.
+   * @param restServerState the {@link RestServerState} object to use.
+   * @param sslFactory the {@link SSLFactory} to use for generating {@link javax.net.ssl.SSLEngine} instances,
+   *                   or {@code null} if SSL is not enabled in this pipeline.
+   */
+  public StorageServerNettyChannelInitializer(NettyConfig nettyConfig, PerformanceConfig performanceConfig,
+      NettyMetrics nettyMetrics, ConnectionStatsHandler connectionStatsHandler, RestRequestHandler requestHandler,
+      RestServerState restServerState, SSLFactory sslFactory, MetricRegistry metricRegistry) {
+    this.nettyConfig = nettyConfig;
+    this.performanceConfig = performanceConfig;
+    this.nettyMetrics = nettyMetrics;
+    this.connectionStatsHandler = connectionStatsHandler;
+    this.restServerState = restServerState;
+    this.sslFactory = sslFactory;
+    RestRequestMetricsTracker.setDefaults(metricRegistry);
+    this.requestHandler = requestHandler;
+  }
+
+  @Override
+  protected void initChannel(SocketChannel ch) throws Exception {
+    // If channel handler implementations are not annotated with @Sharable, Netty creates a new instance of every class
+    // in the pipeline for every connection.
+    // i.e. if there are a 1000 active connections there will be a 1000 NettyMessageProcessor instances.
+    ChannelPipeline pipeline = ch.pipeline();
+    // connection stats handler to track connection related metrics
+    // if SSL is enabled, add an SslHandler before the HTTP codec
+    if (sslFactory != null) {
+      InetSocketAddress peerAddress = ch.remoteAddress();
+      String peerHost = peerAddress.getHostName();
+      int peerPort = peerAddress.getPort();
+      SslHandler sslHandler = new SslHandler(sslFactory.createSSLEngine(peerHost, peerPort, SSLFactory.Mode.SERVER));
+      pipeline.addLast("SslHandler", sslHandler);
+    }
+    pipeline.addLast(Http2FrameCodecBuilder.forServer().build())
+        .addLast("IdleStateHandler", new IdleStateHandler(0, 0, nettyConfig.nettyServerIdleTimeSeconds))
+        .addLast("Http2MultiplexHandler", new Http2MultiplexHandler(
+            new Http2StreamHandler(nettyMetrics, nettyConfig, performanceConfig, requestHandler)));
```

Comment:
I think we should probably add another handler between IdleStateHandler and Http2MultiplexHandler to deal with IdleStateEvent.

Http2MulitiplexHandler doesn't seem like it will handle IdleStateEvent, so we would probably see all the children channels would be triggered by this event.

Reply:
Got you. Let me think about this.

Reply:
OK to remove IdleStateHandler as discussed offline.